internal/controllers/clusterextension_controller.go

```diff
@@ -116,152 +178,195 @@ func checkForUnexpectedFieldChange(a, b ocv1alpha1.ClusterExtension) bool {
 //
 //nolint:unparam
 func (r *ClusterExtensionReconciler) reconcile(ctx context.Context, ext *ocv1alpha1.ClusterExtension) (ctrl.Result, error) {
-	// gather vars for resolution
-	vars, err := r.variables(ctx)
-	if err != nil {
-		ext.Status.InstalledBundle = nil
-		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted due to failure to gather data for resolution", ext.GetGeneration())
-		ext.Status.ResolvedBundle = nil
-		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
-
-		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted due to failure to gather data for resolution", ext.GetGeneration())
-		return ctrl.Result{}, err
-	}
-
+	l := log.FromContext(ctx).WithName("operator-controller")
 	// run resolution
-	selection, err := r.Resolver.Solve(vars)
-	if err != nil {
-		ext.Status.InstalledBundle = nil
-		setInstalledStatusConditionUnknown(&ext.Status.Conditions, "installation has not been attempted as resolution failed", ext.GetGeneration())
-		ext.Status.ResolvedBundle = nil
-		setResolvedStatusConditionFailed(&ext.Status.Conditions, err.Error(), ext.GetGeneration())
-
-		setDeprecationStatusesUnknown(&ext.Status.Conditions, "deprecation checks have not been attempted as resolution failed", ext.GetGeneration())
+	bundle, err := r.resolve(ctx, *ext)
+	if err != nil || bundle == nil {
+		l.V(1).Info("bundle resolve error: %v\n", err)
+
+		errorMessage := err.Error()
+		packageNotFoundPattern := regexp.MustCompile(`^no package "([^"]+)" found$`)
+		versionNotFoundPattern := regexp.MustCompile(`^no package "([^"]+)" matching version "([^"]+)" found$`)
+
+		switch {
+		case packageNotFoundPattern.MatchString(errorMessage):
+			setResolvedStatusConditionFailed(&ext.Status.Conditions, errorMessage, ext.Generation)
+			ensureAllConditionsWithReason(ext, "ResolutionFailed", errorMessage)
+
+		case versionNotFoundPattern.MatchString(errorMessage):
+			setInstalledStatusConditionUnknown(&ext.Status.Conditions, errorMessage, ext.Generation)
```

Comment:
Also, `versionNotFound` alone - has the installedStatus set to unknown, whereas others don't. Can we keep it consistent?

Reply:
Well we inconsistently test for it then:

https://github.com/operator-framework/operator-controller/blob/c6d6f07881bf8aa6a8de0dfaf5a48a5968c1a7ca/internal/controllers/clusterextension_controller_test.go#L137

Want it to be Unknown for Version missing (TestClusterExtensionNonExistentVersion) but don't when it's the Package missing (TestClusterExtensionNonExistentPackage). Should it always be Unknown?